### PR TITLE
Config refactoring

### PR DESCRIPTION
**Pull request by marians**
In this PR I'm introducing viper as a means to deal with settings to be set via defaults, config file, environment variables, and flags.

Changes affecting users:

- The config path changes from `$HOME/.gsctl` to `$HOME/.config/gsctl` (https://github.com/giantswarm/gsctl/issues/39), to avoid clutter of a user's home dir. Existing config is migrated automatically.

Additionally:
- unit tests for the `config` package are added.
- tests no longer drop certificates in the user's actual certs dir (towards #56)

More fruits of this will hopefully be harvested in future PRs.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -174,19 +222,26 @@ func getKubeconfigPaths(homeDir string) []string {
...
+	}
+	apiClient := client.NewClient(clientConfig)
+
 	authHeader := "giantswarm " + Config.Token
-	orgsResponse, _, err := client.GetUserOrganizations(authHeader, requestIDHeader, activityName, cmdLine)
+	orgsResponse, _, err := apiClient.GetUserOrganizations(authHeader, requestIDHeader, activityName, cmdLine)
 	if err != nil {
 		return "", err
 	}
 	if orgsResponse.StatusCode == 10000 {
```
Comment (rossf7) [inline config/config.go]: Why is this status code 10000? I'm confused about what's happening here.

File: config/config.go [hunk 2]
```diff
@@ -202,3 +257,26 @@ func GetDefaultCluster(requestIDHeader, activityName, cmdLine, cmdAPIEndpoint st
...
+	}
+
+	_, err = os.Stat(LegacyConfigDirPath)
+	if os.IsNotExist(err) {
+		// old config dir does not exist
+		return nil
+	}
+
+	// ensure ~/.config exists
+	os.MkdirAll(path.Dir(DefaultConfigDirPath), 0700)
```
Comment (rossf7) [inline config/config.go]: Bit of a nitpick but os.MkdirAll can return an error. Should be OK as its the home dir but might be good to check for it.

File: config/config_test.go [hunk 3]
```diff
@@ -1,8 +1,135 @@
...
+	dir := tempDir()
+	defer os.RemoveAll(dir)
+	filePath := path.Join(dir, ConfigFileName+"."+ConfigFileType)
+
+	// write dummy config
+	file, fileErr := os.Create(filePath)
+	if fileErr != nil {
+		t.Error(fileErr)
+	}
+	file.WriteString("email: email@example.com\ntoken: some-token\n")
```
Comment (rossf7) [inline config/config_test.go]: The tests are great but I'd prefer it the email and token were vars as they are used multiple times.

File: config/config.go (hunk 1 above)
Comment (marians) [inline config/config.go]: That's our legacy API stuff. See http://firstgen-docs.giantswarm.io/reference/api/v1/#GetMemberships

We are using that v1 functions still, as it hasn't been migrated to v4 yet. Shame.

File: config/config.go (hunk 2 above)
Comment (marians) [inline config/config.go]: Right. Changed.

File: config/config_test.go (hunk 3 above)
Comment (marians) [inline config/config_test.go]: Changed

File: config/config.go (hunk 1 above)
Comment (rossf7) [inline config/config.go]: Ah OK I was getting confused between status code and response code. Cool it can be taken care of with the v4 migration.

**Comment by marians** [comment]
@rossf7 Ready for another check

**Comment by marians** [comment]
Thanks for the review!

**Pull request merged**